Issue #14424: Add option to skip validation based on list of annotations

Lmh-java opened the pull request:
Feature for Issue #14424 

Diff Regression config: https://gist.githubusercontent.com/Lmh-java/814de3d973b95b957463eaab2c83f39b/raw/5f3e60d41571b0bff3e6f344af0f7e4ea7a0f5bd/base-util-constructor-config.xml
Diff Regression patch config: https://gist.githubusercontent.com/Lmh-java/ca0caa46b1f782eaf967eddd917cbda7/raw/c3090400a908dce63cb1ac792462fe1d102dac74/patch-util-constructor-config.xml
Diff Regression projects: https://gist.githubusercontent.com/Lmh-java/995679903bfdee2f3b1a8ab2e1578493/raw/068a288ff3424ad4f8c48e8667f1aeef32a6c825/skip-annotation-projects.properties

Lmh-java commented:
GitHub, generate report

Lmh-java commented:
GitHub, generate report

Lmh-java commented:
@nrmancuso @romani Please take a look

nrmancuso commented:
Github, generate site

romani commented on src/main/java/com/puppycrawl/tools/checkstyle/checks/design/HideUtilityClassConstructorCheck.java:
```diff
@@ -91,8 +108,9 @@ public int[] getRequiredTokens() {
 
     @Override
     public void visitToken(DetailAST ast) {
+        // classes with specific annotations should be skipped
```
please remove this, it is written in code by `shouldIgnoreClass(ast)`

romani reviewed (changes requested):
only one minor:

romani commented on src/xdocs-examples/resources/com/puppycrawl/tools/checkstyle/checks/design/hideutilityclassconstructor/Example2.java:
```diff
@@ -0,0 +1,48 @@
+}
+
+@interface SkipMe {}
+// xdoc section -- end
```
move this comment up to hide annotation declaration, in example it does not matter.

romani commented on src/xdocs-examples/resources/com/puppycrawl/tools/checkstyle/checks/design/hideutilityclassconstructor/Example2.java:
```diff
@@ -0,0 +1,48 @@
+}
+
+@SkipMe
+class IgnoredClass1 { // ok, ignored by annotation
```
lets change `SkipMe` to `SpringBootApplication` to let user see exact example and it will help easily understand.

romani commented on src/xdocs-examples/resources/com/puppycrawl/tools/checkstyle/checks/design/hideutilityclassconstructor/Example2.java:
```diff
@@ -0,0 +1,48 @@
+}
+
+@java.lang.Deprecated
+class IgnoredClass2 { // ok, ignored by annotation
```
please update Example1 with such annotations. All Examples should be the same by code
See screenshorts at https://github.com/checkstyle/checkstyle/issues/13345

romani reviewed (changes requested):
items:

romani commented:
@Lmh-java, please reply "done" to each review item, to give us clear singnal  your arer ready for review and nothing is missed to be addressed. Do not resolve conversations, maintainers will do this as confirmation that is done as expected.

romani commented on src/xdocs/checks/design/hideutilityclassconstructor.xml.template:
```diff
@@ -57,6 +66,25 @@ public class StringUtils // not final to allow subclassing
+        </macro>
+        <p id="Example2-code">Example:</p>
+        <macro name="example">
+          <param name="path"
+                   value="resources/com/puppycrawl/tools/checkstyle/checks/design/hideutilityclassconstructor/Example2.java"/>
+            <param name="type" value="code"/>
+        </macro>
+        <p>
+          Note: Annotation names specified in the <code>ignoreAnnotatedBy</code> property
+          must be an exact match for the annotation name on the class.
```
better to put this note to property description.

romani commented on src/xdocs-examples/resources/com/puppycrawl/tools/checkstyle/checks/design/hideutilityclassconstructor/Example1.java:
```diff
@@ -34,7 +35,8 @@ protected Bar() {
 
-class UtilityClass { // violation
+@SpringBootApplication // violation
+class UtilityClass {
```
ApplicationClass, please rename others of same name.

romani commented on src/main/java/com/puppycrawl/tools/checkstyle/checks/design/HideUtilityClassConstructorCheck.java:
```diff
@@ -142,6 +159,26 @@ private static boolean isStatic(DetailAST ast) {
     }
 
+    /**
+     * Setter to ignore classes annotated with the specified annotation(s).
```
please add `Name in this property must be an exact match for the annotation name on the class. If target class has annotation specified as fully qualified name (with package) and by simple name, this property also need to have two values: simple name and fully qualified name.`

romani reviewed (changes requested):
last minors:

romani commented:
@Lmh-java , please reply "done" to each review item, it help a lot.

romani commented on src/main/java/com/puppycrawl/tools/checkstyle/checks/design/HideUtilityClassConstructorCheck.java:
```diff
@@ -51,6 +55,18 @@
+ * with the specified annotation(s). Name in this property must
+ * be an exact match for the annotation name on the class. If target
+ * class has annotation in fully qualified name (with package), this
+ * property should too. If target class has annotation in simple name,
```
`, this property should have such name too.`

romani reviewed (changes requested):
last:

romani commented:
GitHub, generate website

romani commented on src/xdocs-examples/resources/com/puppycrawl/tools/checkstyle/checks/design/hideutilityclassconstructor/Example1.java:
```diff
@@ -18,23 +19,24 @@ public static void fun() {
 }
 
-class Foo { // OK
+class Foo1 { // OK
```
please remove all '"// OK", no violation means ok, we missed it in previos updates

romani reviewed (changes requested):
items:

Lmh-java commented:
> @Lmh-java , please reply "done" to each review item, it help a lot.

Yes, I believe I have replied "done" to each review item after you suggested it from last time. Did I do something incorrectly so you cannot see it.

romani commented:
Please reply "done" as confirmation that get best practice of PR review

Lmh-java commented:
> Please reply "done" as confirmation that get best practice of PR review

<img width="824" alt="Snipaste_2024-03-28_13-21-53" src="https://github.com/checkstyle/checkstyle/assets/52687312/73300c51-6ffe-4dce-9a67-d19becea4c2a">

That's wired, I have always been replying with "done" since you suggested. Can you see them?

romani reviewed (approved):
ok to merge

nrmancuso commented on src/main/java/com/puppycrawl/tools/checkstyle/checks/design/HideUtilityClassConstructorCheck.java:
```diff
@@ -51,6 +55,18 @@
  * </pre>
+ * <ul>
+ * <li>
+ * Property {@code ignoreAnnotatedBy} - Ignore classes annotated
+ * with the specified annotation(s). Name in this property must
+ * be an exact match for the annotation name on the class. If target
+ * class has annotation in fully qualified name (with package), this
+ * property should have such name too. If target class has annotation
+ * in simple name, this property needs to have the same simple name.
```
```suggestion
 * Property {@code ignoreAnnotatedBy} - Ignore classes annotated
 * with the specified annotation(s). Annotation names provided in this property
 * must exactly match the annotation names on the classes. If the target class has annotations
 * specified with their fully qualified names (including package), the annotations in this
 * property should also be specified with their fully qualified names. Similarly, if the target
 * class has annotations specified with their simple names, this property should contain the
 * annotations with the same simple names.
```

Lmh-java replied:
done, and rebased on master

nrmancuso commented on src/main/java/com/puppycrawl/tools/checkstyle/checks/design/HideUtilityClassConstructorCheck.java:
```diff
@@ -74,6 +90,15 @@ public class HideUtilityClassConstructorCheck extends AbstractCheck {
      */
     public static final String MSG_KEY = "hide.utility.class";
 
+    /**
+     * Ignore classes annotated with the specified annotation(s). Name in
+     * this property must be an exact match for the annotation name on the class.
+     * If target class has annotation in fully qualified name (with package), this
+     * property should have such name too. If target class has annotation in simple
+     * name, this property needs to have the same simple name.
+     */
```
This will need to be updated to reflect the changes above.

Lmh-java replied:
done

nrmancuso reviewed (changes requested):
Minor documentation update:

Lmh-java commented:
@nrmancuso ping :)

nrmancuso commented:
Rebasing on master, OK to merge as CI passes

rnveach commented on src/main/java/com/puppycrawl/tools/checkstyle/checks/design/HideUtilityClassConstructorCheck.java:
```diff
@@ -142,6 +170,32 @@ private static boolean isStatic(DetailAST ast) {
+     * @param annotationNames specified annotation(s)
+     * @since 10.15.0
+     */
+    public void setIgnoreAnnotatedBy(String... annotationNames) {
```
Move setter to before `@Override` methods.

rnveach commented on src/xdocs/checks/design/hideutilityclassconstructor.xml.template:
```diff
@@ -57,6 +66,21 @@ public class StringUtils // not final to allow subclassing
         </macro>
+
+        <p id="Example2-config">
+          To configure the check:
```
This verbiage needs to change to specify what kind of configuration we are doing here. This is the exact same verbiage as Example 1, so if someone sees these 2 they will question why didn't the first example already show how to configure and what makes Example 2 any different. Look at our other documentation examples.

Lmh-java replied:
done

rnveach replied:
I don't see how this is done. Example 2 still just says `To configure the check`. I am expecting something like `To configure the check to ...`.

rnveach reviewed (changes requested):
> https://checkstyle-diff-reports.s3.us-east-2.amazonaws.com/f763b2e_2024052758/reports/diff/index.html

This only shows regression. We need a report for the new property as well. Configure it for some annotation. You can use the annotation specified in the issue.

Lmh-java commented:
> > https://checkstyle-diff-reports.s3.us-east-2.amazonaws.com/f763b2e_2024052758/reports/diff/index.html
> 
> This only shows regression. We need a report for the new property as well. Configure it for some annotation. You can use the annotation specified in the issue.

The regression report is here.

https://checkstyle-diff-reports.s3.us-east-2.amazonaws.com/75e5ab1_2024052933/reports/diff/index.html

I have to use a different project to test this new feature. I have gone through our default regression project list, but didn't see any valid test case satisfying all the conditions here.

rnveach commented:
> gone through our default regression project list, but didn't see any valid test case 

Report is fine, just that if original regression showed nothing for this change I would still leave the projects in there as proof they don't show anything.

Lmh-java commented:
GitHub, generate report

Lmh-java commented:
https://github.com/checkstyle/checkstyle/pull/14701#issuecomment-2014384416
https://github.com/checkstyle/checkstyle/pull/14701#issuecomment-2014471626
These two reports have shown there is nothing changed. But FYI, I will aggregate the project list for clearer demonstration.

Lmh-java commented:
GitHub, generate report

rnveach commented:
> I noticed that we are not using local-checkstyle|local| config in the default project list. And if I use that, there will be some random differences.
> If I run locally to compare between these two configs, there are no those difference as shown in the report. 
> local-checkstyle|local|../../../|master

It depends on where this local is coming from and if anything else is using it. If this is the same git repository provided to regression as the basis for checkstyle, then you have to know that part of the work regression does to do its magic is to switch branches in that git repository. It has to switch between master and the PR to install both versions of checkstyle in order to run them. That switching may cause issues with you defining it as a property to test on as it is only expecting to stay in master and not move around.

This is an issue to be able to set up a way to do regression on the changes in the PR itself.

rnveach commented on src/xdocs-examples/resources/com/puppycrawl/tools/checkstyle/checks/design/hideutilityclassconstructor/Example2.java:
```diff
@@ -0,0 +1,47 @@
+  <module name="TreeWalker">
+    <module name="HideUtilityClassConstructor">
+      <property name="ignoreAnnotatedBy"
+        value="SpringBootApplication, java.lang.Deprecated" />
```
> Annotation names provided in this property must exactly match the annotation names on the classes. If the target class has annotations specified with their fully qualified names (including package), the annotations in this property should also be specified with their fully qualified names. Similarly, if the target class has annotations specified with their simple names, this property should contain the annotations with the same simple names.

Show some of these issues in your example.

I assume `Deprecated` by itself won't work. Show that and put a comment to that effect.

rnveach replied:
Can you also provide me where we said we would be so strict?

Lmh-java replied:
> I assume Deprecated by itself won't work.

Only `Deprecated` will not work if it doesn't match the annotation used.

Lmh-java replied:
> Can you also provide me where we said we would be so strict?

We mentioned this at https://github.com/checkstyle/checkstyle/issues/14424#issuecomment-2008610998. We agreed to do exact match first as https://github.com/checkstyle/checkstyle/pull/14553. We might do pattern matching later, but they need to solve the problem of https://github.com/checkstyle/checkstyle/issues/5434 first.

Lmh-java replied:
> > Annotation names provided in this property must exactly match the annotation names on the classes. If the target class has annotations specified with their fully qualified names (including package), the annotations in this property should also be specified with their fully qualified names. Similarly, if the target class has annotations specified with their simple names, this property should contain the annotations with the same simple names.
> 
> Show some of these issues in your example.
> 
> I assume `Deprecated` by itself won't work. Show that and put a comment to that effect.

done

rnveach replied:
This does not look done. I don't see any examples were we fail to match because partials was only provided.

It is also not documented that we understand `java.lang` when we say `property must exactly match`. `Deprecated` is not an exact match for `java.lang.Deprecated`.

rnveach replied:
I removed my comment here and added some more below.

rnveach commented on src/xdocs/checks/design/hideutilityclassconstructor.xml:
```diff
@@ -42,6 +42,27 @@ public class StringUtils // not final to allow subclassing
+              <td>Ignore classes annotated with the specified annotation(s). Annotation names provided in this property must exactly match the annotation names on the classes. If the target class has annotations specified with their fully qualified names (including package), the annotations in this property should also be specified with their fully qualified names. Similarly, if the target class  ...
+              <td><a href="../../property_types.html#String.5B.5D">String[]</a></td>
+              <td><code>{}</code></td>
+              <td>10.16.0</td>
```
Version needs updated.

romani replied:
should be 10.17.0

Lmh-java replied:
done

rnveach reviewed (changes requested):
Last things on documentation.

romani commented:
@

romani commented:
@Lmh-java , please help us to finish this PR, we are so close to merge state.

Lmh-java commented:
> @Lmh-java , please help us to finish this PR, we are so close to merge state.
> looks like only version should be changed.

Sorry, I was quite occupied with other stuffs.

I have updated a new example for https://github.com/checkstyle/checkstyle/pull/14701#discussion_r1583561783 and updated the version.

Lmh-java commented:
GitHub, generate site

rnveach commented on src/xdocs/checks/design/hideutilityclassconstructor.xml:
```diff
@@ -42,6 +42,27 @@ public class StringUtils // not final to allow subclassing
+              <td>Ignore classes annotated with the specified annotation(s). Annotation names provided in this property must exactly match the annotation names on the classes. If the target class has annotations specified with their fully qualified names (including package), the annotations in this property should also be specified with their fully qualified names. Similarly, if the target class  ...
+              <td><a href="../../property_types.html#String.5B.5D">String[]</a></td>
+              <td><code>{}</code></td>
+              <td>10.17.0</td>
```
@romani @nrmancuso I think we have an issue on property versions. Another PR I looked at had 2 digit version, this one has 3. I looked at https://checkstyle.org/checks/annotation/annotationlocation.html#Properties and it has 2 digits.

We are missing some validation here and we need to correct it.

@Lmh-java Please create a new issue for this. We need to fix existing cases as well as add proper validation that everything is 3 digits if there isn't, imo.

romani replied:
We use to have 2 digits release numbers. Only in 2023 we started doing 3 digits. Suppression list will be pretty long.

rnveach replied:
I am resolving this.

rnveach commented on src/xdocs-examples/resources/com/puppycrawl/tools/checkstyle/checks/design/hideutilityclassconstructor/Example3.java:
```diff
@@ -0,0 +1,46 @@
+package com.puppycrawl.tools.checkstyle.checks.design.hideutilityclassconstructor;
+
+// xdoc section -- start
+@java.lang.Deprecated  // violation
```
add some verbiage that this isn't an exact match because of the package name.

rnveach commented on src/xdocs-examples/resources/com/puppycrawl/tools/checkstyle/checks/design/hideutilityclassconstructor/Example3.java:
```diff
@@ -0,0 +1,46 @@
+</module>
+*/
+
+package com.puppycrawl.tools.checkstyle.checks.design.hideutilityclassconstructor;
```
I would much rather this be combined with example 2. It makes more sense to see side by side, exact match suppresses, non-exact match is a violation.

rnveach commented on src/xdocs-examples/resources/com/puppycrawl/tools/checkstyle/checks/design/hideutilityclassconstructor/Example3.java:
```diff
@@ -0,0 +1,46 @@
+}
+
+@SpringBootApplication
+class ApplicationClass3 { // OK, skipped by annotation
```
Same when this becomes a violation.

romani commented:
@Lmh-java , do you have time to finish update we 99% done.

romani commented:
@kkoutsilis, it would be awesome if you help us to finish this update.
It is abandoned. Please copy code to your PR and please finish it. I will close this PR as soon as you open new.

Thanks a lot in advance.

kkoutsilis commented:
> @kkoutsilis, it would be awesome if you help us to finish this update. It is abandoned. Please copy code to your PR and please finish it. I will close this PR as soon as you open new.
> 
> Thanks a lot in advance.

@romani I will have a look when I have some free time, thanks!
